proposals/module-types/Explainer.md

```diff
@@ -602,32 +568,76 @@ Therefore, module and instance types can appear in both the imports and exports
 of module types and instance types.
 
 
+### Binary Format Considerations
+
+This proposal defines two new [sections] in the binary format: Module,
+ModuleCode and Instance. Their proposed ordering of sections is:
+1. Type Section
+2. Import Section
+3. Module Section
+4. Instance Section
+5. Function Section
+6. ...
+7. ModuleCode Section
+8. Code Section
+9. Data Section
+
+The distinction between the Module and ModuleCode Sections is the same as the
+distinction between the Function and Code Sections: the Module Section declares
+the module types of each nested module while the ModuleCode Section provides the
+definition of each module. Module definitions in the ModuleCode Section are
+decoded with the same [`module` binary format production] that decodes top-level
+modules, making `module` a recursive production. As with functions, splitting
+module types from definitions allows all modules to be decoded, validated and
+optimized in parallel.
+
+This section ordering is implied by:
+* Module definitions depend on the Type Section and, later, with [Type Imports],
+  the Import Section.
+* Instance definitions (in the Instance Section) depend on module types (in the
+  Module Section).
+* The exports of instance definitions in the Instance Section can be used in
+  all succeeding sections (even, once instances can [export types], the Function
+  Section).
+* Since a module is instantiated and executed (via its `start` function) after
+  all its nested modules, having Code after ModuleCode allows Code compilation
+  to potentially better overlap ModuleCode execution in streaming scenarios.
+
+There are several kinds of `alias` definitions introduced above; they go in
+different sections:
+* Aliases of parent modules' types go in the Type Section as a new kind of
+  type definition.
+* Aliases of parent modules' modules go in the Module Section as a new kind of
+  module definition.
+* Aliases of imported or nested instances' exports go in the Instance Section
+  as a new kind of definition (of the aliased export's type).
```

Comment:
This leaves a serious hole: a type definition cannot refer to types from an imported or nested instance. At least with nominal types, you will need this. (With structural you would be able to get away without it.)

I think we'd need a way for type definitions to (1) refer to local type aliases, and (b) for those type aliases to refer to local instance aliases (to access more deeply nested types).

Reply:
Eww.  I see your point.  So already I was realizing that, with Type Imports, there's a sortof cyclic relationship between the Type and Imports Sections.  I think what that means is that it's necessary to decode and validate the Type+Import section as a whole unit.  If we want typedefs to include instance type exports, then I suppose it makes sense to expand that cyclic set to the {Type,Import,Nested} sections?  Preventing bad cycles no longer seems trivial, though...  Is that what you're thinking?

Reply:
Yeah. It is rather unfortunate. The semantics of module/instance stuff depends on declarations being ordered, and with cyclic dependencies between sections you lose a big part of that information and have to reconstruct it, which may require a proper dependency analysis in the decoder/validator. :(  Not very appealing. The alternative would be allowing multiple/interleaved occurrences of sections, which I reckon to be simpler ultimately.

Reply:
Which sections would need be repeated/interleaved? Just `type`, `import`, `module`, right?

Reply:
And `instance` and `alias`, too, assuming those are separate sections.

Reply:
Ok, yeah, it does seem simpler to allow interleaved sections if that means we can just use index ordering to enforce acyclicy.  Just to confirm: while there *can* be cycles between typedefs and type imports (via their subtype bounds), there *can't* be any cycles that go through modules, instances or aliases, right?  Thus, we can simply append definitions to index spaces in binary format order and just check indices relative to the current size of the index spaces up to the point of the definition, right?

(Btw, I was thinking it would be nice if type cycles could be confined be only within a single section so validation could essentially proceed section-by-section.  To deal with the typedef/type-import cycles, I was thinking maybe "type imports" could actually go inside the type section as a new kind of typedef.  Would that work?)

Reply:
Yes, if we were to allow type cycles through instances then we would be entering the realm of recursive modules. =:-O That'd be a whole new level of complexity, best avoided as long as we possibly can, and a feature which, to be supported properly, would require additional extensions.

Reply:
Ok, that's what I was hoping :)  I just pushed a commit that proposes allowing the repeating/interleaving.  I'll close this, but feel free to reopen or open new discussions on the new proposed text.